### ISSUE TITLE
[BUG] 由于某些原因未设置 CUDA_DEVICE_MAX_CONNECTIONS环境变量，loss 下降很慢

### ISSUE THREAD

**Issue by luohuan02** [QwenLM/Qwen#577]
### 是否已有关于该错误的issue或讨论？ | Is there an existing issue / discussion for this?

- [X] 我已经搜索过已有的issues和讨论 | I have searched the existing issues / discussions

### 该问题是否在FAQ中有解答？ | Is there an existing answer for this in FAQ?

- [X] 我已经搜索过FAQ | I have searched FAQ

### 当前行为 | Current Behavior

训练时，不设置 CUDA_DEVICE_MAX_CONNECTIONS=1 环境变量，loss 从 3+ 降到 1+，如果设置了CUDA_DEVICE_MAX_CONNECTIONS=1  环境变量，loss  从 3+ 降到 0.1+ ，下降快很多，请教一下这个是什么原因呢？

### 运行环境 | Environment

```Markdown
- OS:
- Python:
- Transformers:
- PyTorch:
- CUDA (`python -c 'import torch; print(torch.version.cuda)'`):
```

**Comment by luohuan02** [comment]
上面可能描述格式不太合适，具体描述如下，由于某些原因，在训练时，忘记设置 CUDA_DEVICE_MAX_CONNECTIONS=1 环境变量，训练 10 轮 loss 从 3+ 只降到 1+，后面参考设置了CUDA_DEVICE_MAX_CONNECTIONS=1 环境变量，loss 从 3+ 降到 0.1+ ， loss 下降快很多，有没有大佬知道这个环境变量为什么会产生这种影响呢？没有设置环境变量还会有其它什么问题么？

**Comment by jklj077** [comment]
您好, 该环境变量大多数情况下仅影响并行性，不影响正确性（对于影响正确性的，正常会直接报错退出）。
1. 请先尝试下仅去除该环境变量，是否仍然有loss下降慢的问题，以排除其它因素（包括随机因素）的影响。
2. 请提供下相关的运行细节（使用的哪个finetune脚本，各项配置情况，使用了多少张卡等）

**Comment by luohuan02** [comment]
> 您好, 该环境变量大多数情况下仅影响并行性，不影响正确性（对于影响正确性的，正常会直接报错退出）。
> 
> 1. 请先尝试下仅去除该环境变量，是否仍然有loss下降慢的问题，以排除其它因素（包括随机因素）的影响。
> 2. 请提供下相关的运行细节（使用的哪个finetune脚本，各项配置情况，使用了多少张卡等）

不好意思，我是在其他项目兼容 qwen 代码的时候出现的上述问题，使用 qwen 脚本和代码没有复现出来，我在看看有什么区别，谢谢了，后续结果我会同步到这里

**Comment by luohuan02** [comment]
> 您好, 该环境变量大多数情况下仅影响并行性，不影响正确性（对于影响正确性的，正常会直接报错退出）。
> 
> 1. 请先尝试下仅去除该环境变量，是否仍然有loss下降慢的问题，以排除其它因素（包括随机因素）的影响。
> 2. 请提供下相关的运行细节（使用的哪个finetune脚本，各项配置情况，使用了多少张卡等）

你好，最近两天补充了一些实验，结论如下：CUDA_DEVICE_MAX_CONNECTIONS 和 batch padding 影响 loss 速度
1、如果设置 CUDA_DEVICE_MAX_CONNECTIONS=1 无论是否 padding 到 max（比如4096），loss 下降正常
2、如果不设置 CUDA_DEVICE_MAX_CONNECTIONS=1 并且 padding 到 max，loss 下降正常
3、如果不设置 CUDA_DEVICE_MAX_CONNECTIONS=1，只是做 batch padding（非全局 padding），loss 下降非常慢

动态 padding 方式具体 finetune.py 修改如下两处，启动脚本参数完全参考 sft 参数，只是环境变量设置不同：
1、process 修改（这里是想做预训练，所以如下修改）
```
def preprocess(
    sources,
    tokenizer: transformers.PreTrainedTokenizer,
    max_len: int
) -> Dict:
    input_ids = []
    targets = []
    # 动态padding
    attention_mask = []
    for i, source in enumerate(sources):
        tokenized = tokenizer(source)
        input_id = tokenized.input_ids + [tokenizer.eod_id]
        target = input_id.copy()
        att_mask = tokenized.attention_mask
        input_ids.append(input_id[:max_len])
        targets.append(target[:max_len])
        attention_mask.append(att_mask[:max_len])

    return dict(
        input_ids=input_ids,
        labels=targets,
        attention_mask=attention_mask,
    )
```

2、Trainer 增加 transformers.DataCollatorForTokenClassification, 参考如下
```
    trainer = Trainer(
        model=model, tokenizer=tokenizer, args=training_args, **data_module,
        data_collator=transformers.DataCollatorForTokenClassification(tokenizer=tokenizer, padding="longest",
                                                                      max_length=training_args.model_max_length,
                                                                      label_pad_token_id=-100),
    )
```
